File: bitlap-core/src/main/kotlin/org/bitlap/core/sql/rule/BitlapRelConverter.kt
/**
 * Copyright (C) 2023 bitlap.org .
 */
package org.bitlap.core.sql.rule

import org.apache.calcite.plan.RelOptRuleCall
import org.apache.calcite.plan.hep.HepRelVertex
import org.apache.calcite.rel.RelNode
import org.apache.calcite.rel.SingleRel
import org.apache.calcite.rel.logical.LogicalAggregate
import org.apache.calcite.rel.logical.LogicalFilter
import org.apache.calcite.rel.logical.LogicalProject
import org.apache.calcite.rel.logical.LogicalSort
import org.apache.calcite.rel.logical.LogicalTableScan
import org.apache.calcite.rel.logical.LogicalUnion
import org.apache.calcite.rel.logical.LogicalValues
import org.bitlap.core.sql.rel.BitlapAggregate
import org.bitlap.core.sql.rel.BitlapFilter
import org.bitlap.core.sql.rel.BitlapNode
import org.bitlap.core.sql.rel.BitlapProject
import org.bitlap.core.sql.rel.BitlapSort
import org.bitlap.core.sql.rel.BitlapTableScan
import org.bitlap.core.sql.rel.BitlapUnion
import org.bitlap.core.sql.rel.BitlapValues
import org.joor.Reflect

/**
 * Convert calcite rel node to bitlap rel node.
 */
class BitlapRelConverter : AbsRelRule(RelNode::class.java, "BitlapRelConverter") {

    override fun convert0(rel: RelNode, call: RelOptRuleCall): RelNode {
        val root = call.planner.root.clean()
        // only convert root rel node
        if (rel != root) {
            return rel
        }
        return this.convert00(rel, call)
    }

    private fun convert00(rel: RelNode, call: RelOptRuleCall): RelNode {
        return when (rel) {
            // has been converted
            is BitlapNode -> {
                rel
            }
            is HepRelVertex -> {
                // next parent should be HepRelVertex's parent and next level should be current level, because it's a wrapper
                rel.also {
                    Reflect.on(it).call("replaceRel", this.convert00(it.currentRel, call))
                }
            }
            is LogicalSort -> {
                this.convert00(rel.input, call).injectParent {
                    BitlapSort(rel.cluster, rel.traitSet, it, rel.collation, rel.offset, rel.fetch)
                }
            }
            is LogicalAggregate -> {
                this.convert00(rel.input, call).injectParent {
                    BitlapAggregate(
                        rel.cluster, rel.traitSet, rel.hints, it,
                        rel.groupSet, rel.groupSets, rel.aggCallList
                    )
                }
            }
            is LogicalProject -> {
                this.convert00(rel.input, call).injectParent {
                    BitlapProject(rel.cluster, rel.traitSet, rel.hints, it, rel.projects, rel.rowType)
                }
            }
            is LogicalFilter -> {
                this.convert00(rel.input, call).injectParent {
                    BitlapFilter(rel.cluster, rel.traitSet, it, rel.condition, rel.variablesSet)
                }
            }
            is LogicalUnion -> {
                val union = BitlapUnion(rel.cluster, rel.traitSet, rel.inputs, rel.all)
                rel.inputs.map {
                    this.convert00(it, call).injectParent {
                        union
                    }
                }.first()
            }
            is LogicalTableScan -> {
                BitlapTableScan(rel.cluster, rel.traitSet, rel.hints, rel.table)
            }
            is LogicalValues -> {
                BitlapValues(rel.cluster, rel.traitSet, rel.rowType, rel.tuples)
            }
            else -> {
                when (rel) {
                    is SingleRel -> {
                        this.convert00(rel.input, call).injectParent {
                            rel.replaceInput(0, it)
                            rel
                        }
                    }
                    else -> {
                        throw IllegalArgumentException("Invalid converted rel node: ${rel.digest}")
                    }
                }
            }
        }
    }
}

File: bitlap-core/src/main/scala/org/bitlap/core/sql/rule/BitlapRelConverter.scala
/**
 * Copyright (C) 2023 bitlap.org .
 */
package org.bitlap.core.sql.rule

import scala.jdk.CollectionConverters._

import org.bitlap.core.extension._
import org.bitlap.core.sql.rel.BitlapAggregate
import org.bitlap.core.sql.rel.BitlapFilter
import org.bitlap.core.sql.rel.BitlapNode
import org.bitlap.core.sql.rel.BitlapProject
import org.bitlap.core.sql.rel.BitlapSort
import org.bitlap.core.sql.rel.BitlapTableScan
import org.bitlap.core.sql.rel.BitlapUnion
import org.bitlap.core.sql.rel.BitlapValues

import org.apache.calcite.plan.RelOptRuleCall
import org.apache.calcite.plan.hep.HepRelVertex
import org.apache.calcite.rel.RelNode
import org.apache.calcite.rel.SingleRel
import org.apache.calcite.rel.logical.LogicalAggregate
import org.apache.calcite.rel.logical.LogicalFilter
import org.apache.calcite.rel.logical.LogicalProject
import org.apache.calcite.rel.logical.LogicalSort
import org.apache.calcite.rel.logical.LogicalTableScan
import org.apache.calcite.rel.logical.LogicalUnion
import org.apache.calcite.rel.logical.LogicalValues
import org.joor.Reflect

/** Convert calcite rel node to bitlap rel node.
 */
class BitlapRelConverter extends AbsRelRule(classOf[RelNode], "BitlapRelConverter") {

  override def convert0(rel: RelNode, call: RelOptRuleCall): RelNode = {
    val root = call.getPlanner.getRoot.clean()
    // only convert root rel node
    if (rel != root) {
      return rel
    }
    this.convert00(rel, call)
  }

  private def convert00(_rel: RelNode, call: RelOptRuleCall): RelNode = {
    _rel match {
      // has been converted
      case rel: BitlapNode =>
        rel
      case rel: HepRelVertex =>
        // next parent should be HepRelVertex's parent and next level should be current level, because it's a wrapper
        rel.also { it =>
          Reflect.on(it).call("replaceRel", this.convert00(it.getCurrentRel, call))
        }
      case rel: LogicalSort =>
        this.convert00(rel.getInput, call).injectParent { it =>
          BitlapSort(rel.getCluster, rel.getTraitSet, it, rel.getCollation, rel.offset, rel.fetch)
        }
      case rel: LogicalAggregate =>
        this.convert00(rel.getInput, call).injectParent { it =>
          BitlapAggregate(
            rel.getCluster,
            rel.getTraitSet,
            rel.getHints,
            it,
            rel.getGroupSet,
            rel.getGroupSets,
            rel.getAggCallList
          )
        }
      case rel: LogicalProject =>
        this.convert00(rel.getInput, call).injectParent { it =>
          new BitlapProject(
            rel.getCluster,
            rel.getTraitSet,
            rel.getHints,
            it,
            rel.getProjects,
            rel.getRowType,
            rel.getVariablesSet
          )
        }
      case rel: LogicalFilter =>
        this.convert00(rel.getInput, call).injectParent { it =>
          BitlapFilter(rel.getCluster, rel.getTraitSet, it, rel.getCondition, rel.getVariablesSet)
        }
      case rel: LogicalUnion =>
        val union = BitlapUnion(rel.getCluster, rel.getTraitSet, rel.getInputs, rel.all)
        rel.getInputs.asScala.map { i =>
          this.convert00(i, call).injectParent { _ =>
            union
          }
        }.head
      case rel: LogicalTableScan =>
        BitlapTableScan(rel.getCluster, rel.getTraitSet, rel.getHints, rel.getTable)
      case rel: LogicalValues =>
        BitlapValues(rel.getCluster, rel.getTraitSet, rel.getRowType, rel.getTuples)
      case _ =>
        _rel match {
          case rel: SingleRel =>
            this.convert00(rel.getInput, call).injectParent { it =>
              rel.replaceInput(0, it)
              rel
            }
          case rel =>
            throw IllegalArgumentException(s"Invalid converted rel node: ${rel.getDigest}")
        }
    }
  }
}
